Kubescape self-assessment (#1424)
cncf/tag-security: community/assessments/projects/kubescape/self-assessment.md

  * [Actors](#actors)
  * [Actions](#actions)
  * [Background](#background)
  * [Goals](#goals)
  * [Non-goals](#non-goals)
* [Self-assessment use](#self-assessment-use)
* [Security functions and features](#security-functions-and-features)
* [Project compliance](#project-compliance)
* [Secure development practices](#secure-development-practices)
* [Security issue resolution](#security-issue-resolution)
* [Appendix](#appendix)

## Metadata

|   |  |
| -- | -- |
| Assessment Stage | Incomplete |
| Software |  [Kubescape](https://kubescape.io)  |
| Security Provider | Yes |
| Languages | Go |
| SBOM | Kubescape does not currently generate an SBOM on release|
| Security links | See bellow |

### Security links

| Doc | url |
| -- | -- |
| Reporting security issues | https://github.com/kubescape/project-governance/blob/main/SECURITY.md |

## Overview

Kubescape is an open-source Kubernetes security platform that provides comprehensive security coverage, from left to right across the entire development and deployment lifecycle. It offers hardening, posture management, and runtime security capabilities to ensure robust protection for Kubernetes environments.

### Background

> Kubescape project started from being a CLI scanner testing Kubernetes configurations against the [NSA Hardening Guide](https://www.cisa.gov/news-events/alerts/2022/03/15/updated-kubernetes-hardening-guide). The original creators are [ARMO](https://armosec.io) employees who a started this as a side-project for Kubernetes security posture while their main focus was the workload runtime security.

Reviewer: ```suggestion
Kubescape project started from being a CLI scanner testing Kubernetes configurations against the [NSA Hardening Guide](https://www.cisa.gov/news-events/alerts/2022/03/15/updated-kubernetes-hardening-guide). The original creators are [ARMO](https://armosec.io) employees who started this as a side-project for Kubernetes security posture while their main focus was the workload runtime security.
```

##### Kubescape component

This component is in charge of configuration and host scanning. It is, like the CLI, uses [OPA engine](https://github.com/open-policy-agent/opa) to run the project's own Rego library of rules. It also scans the Kubernetes host to validate their configurations. The output of the scans are stored in the *Storage* component via the Kubernetes API and optionally sent to external API endpoints.

##### KubeVuln component

This component is in charge of the image vulnerability scanning. It can either produce SBOM object in the *Storage* and match the SBOM entries with vulnerabilities, or relies on the *Node agent* to generate SBOM objects on the nodes and then produce vulnerability manfiests and VEX. All the results are stored in the *Storage* component via the Kubernetes API and optionally sent to external API endpoints.

##### Node agent component

This component has multiple purposes all bound to information available on Kubernetes nodes.

> * Produces SBOMs from the images avialable on the node (used by *KubeVuln*)

Reviewer: ```suggestion
* Produces SBOMs from the images available on the node (used by *KubeVuln*)
```

* Creates *NetworkNeighborhood* objects using [Inspektor Gadget](https://inspektor-gadget.io) and eBPF. These profiles log the network activity of each container and they stored as objects in the *Storage* component via the Kubernetes API and optionally sent to external API endpoints.
* Monitors container activity via eBPF and evaluates them using its own rule engine that combines static detection rules and anomaly detection to produce alerts that can be exported to AlertManager, Syslog, HTTP endpoints, STDOUT stream and other.

##### Storage component

This component is a Kubernetes [aggregated API extension](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/) service. It stores the different objects produced by the other components and stores them on a volume as files and SQLite. It is a singleton component in the current implementation and cannot be scaled horizontaly, but it is running in 10k node clusters.

##### Synchronizer component

> This component is an optional part of the Kubescape Operator. It enables users to replicate the Kubernetes objects in the cluster (somewhat like `rsynch`) to a remote service. It is used for collecting the Kubescape Operator objects by central services monitoring multiple clusters.

Reviewer: ```suggestion
This component is an optional part of the Kubescape Operator. It enables users to replicate the Kubernetes objects in the cluster (somewhat like `rsync`) to a remote service. It is used for collecting the Kubescape Operator objects by central services monitoring multiple clusters.
```

#### Detailed overview

For a deeper architectural overview go [here](https://github.com/kubescape/helm-charts/blob/doc/charts/kubescape-operator/README.md#in-cluster-components-overview)

#### Objects

* ApplicationProfile - storing workload level information about container activities (file, process, systemcall, capabilities)
* ConfigurationScanSummary - summary object containing results of configuration scanning
* GeneratedNetworkPolicy - object containing proposed network policies for each workload
* KnownServer - list of IP to host conversion
* NetworkNeighborhood - storing workload level information about network activities of each container
* OpenVulnerabilityExchangeContainer - [VEX](https://github.com/openvex) object that was built from the vulnerability scan together with reachability analysis

> * SBOMSyftFiltered - SBOM of running containers filtered by only those that are in use

Reviewer: hopefully this one will disappear next week

Author: Which one?

Reviewer: filtered SBOMs

* SBOMSyft - SBOM of running containers

> * SeccompProfile - Seccomp profile helper object (converting it to JSONs on each host)

Reviewer: ```suggestion
* SeccompProfile - Seccomp profile distribution helper object (converting it to JSONs on each host)
```

* VulnerabilityManifest - vulnerability scan result object
* VulnerabilityManifestSummary - summary object of the vulnerability scans
* WorkloadConfigurationScan -  workload level configuration scan output
* WorkloadConfigurationScanSummary - summary of the workload configuration scan

### Actions

#### Configuration scans

Configuration scans can be triggered in three ways:

> * Continuous scanning: every new or updated object is scanned, this is achieved by "watching" Kubernetes API objects

Reviewer: mention this is still in beta?

Author: I think it is ok

* Scheduled scanning: using pre-configured recurring triggers to scan configurations in the cluster, this is achieved using CronJobs
* Triggered scanning: Kubescape Operator exposing API to trigger scan of specific objects or the whole cluster

#### Vulnerability scans

Vulnerability scans can be triggered in three ways:
* Continuous scanning: every new image that are introduced to the cluster is scanned, this is achieved by "watching" Kubernetes API objects
* Scheduled scanning: using pre-configured recurring triggers to scan for vulnerabilities in the cluster, this is achieved using CronJobs
* Triggered scanning: Kubescape Operator exposing API to trigger scan of specific image and workloads in the cluster

#### Object storage

> In the storage component, full Kubernetes object API is supported with all the CRUD actions.

Reviewer: mention that list only returns metadata for performance reasons?

Author: I don't think it is an important detail in this context

* Synchornizer server: Synchornizer server can trigger the *Synchronizer* component to store objects in Kubernetes API (this is limited by the *ClusterRole* the *Synchronizer* is assigned with)

### Goals

* Kubescape Operator automates scanning and monitoring of Kubernetes clusters
  * provides up to date security posture data
  * supplies proposals of how to improve the security posture of the workload and the cluster
  * provides security observability data
  * alerts on potential breaches to workloads in the cluster
* Kubescape CLI scans pre-deployment and post-deployment configurations and images and provides actionable results

#### Security Goals

* All access to the data produced by Kubescape Operator should be only available to those subjects who has the proper RBAC authorizations
* Kubescape should only share information to Non-Kubernetes endpoints in cases when the user explicitly defined it
* Kubescape should only have access to objects it must need to complete its job
* Triggers should only come from authenticated and authorized sources

### Non-goals

> * Kubescape is not meant to enforce access control beyond what Kubernetes API already provides

Reviewer: what about the webhooks in operator?

Author: WDYM?

Reviewer: we don't provide new RBAC, but we do support validating webhooks in operator

### Communication Channels
  Kubescape Communication Channels

#### Internal Communication
- Slack: Join the Kubescape slack channel in [CNCF slack](https://communityinviter.com/apps/cloud-native/cncf)
- GitHub: Start a [discussion](https://github.com/kubescape/kubescape/discussions) or open an [issue](https://github.com/kubescape/kubescape/issues)
- Security topics: [cncf-kubescape-maintainers@lists.cncf.io](#reporting-security-vulnerabilities)

#### Inbound Communication

Users or prospective users communicate with the Kubescape team through GitHub issues and pull requests. GitHub is a platform that hosts the Kubescape project's codebase and provides features for tracking changes, managing versions, and collaborating on code. Users can report issues, propose changes, or contribute to the project by submitting pull requests.
- GitHub: Start a [discussion](https://github.com/kubescape/kubescape/discussions) or open an [issue](https://github.com/kubescape/kubescape/issues)

#### Outbound Communication
- Twitter: [@Kubescape](https://x.com/Kubescape)

##### Community meetings

> We hold [community meetings](https://zoom.us/j/95174063585) on Zoom, every other week, at 15:00 CET. ([See that in your local time zone](https://time.is/compare/1500_in_CET). :headphones:

Reviewer: issue with headphones icon?

##### Operator component

This component is in charge of commanding and controlling the scans in the cluster. There are multiple configuration options for when and what to scan in the cluster. This component orchestrates these activities by triggering the Kubescape and KubeVuln components.

##### Kubescape component

This component is responsible for  configuration and host scanning. Like the CLI, it uses the [OPA engine](https://github.com/open-policy-agent/opa) to run the project's own Rego library of rules. It also scans the Kubernetes host to validate their configurations. The output of the scans is stored in the *Storage* component via the Kubernetes API and optionally sent to external API endpoints.

##### KubeVuln component

This component is responsible for image vulnerability scanning. It can either produce an SBOM object in the *Storage* and match the SBOM entries with vulnerabilities or rely on the *Node agent* to generate SBOM objects on the nodes and produce vulnerability manifests and VEX documents. All the results are stored in the *Storage* component via the Kubernetes API and optionally sent to external API endpoints.

##### Node agent component

> This component has multiple purposes all bound to information available on Kubernetes nodes.

Reviewer 2: I can't grammatically parse this.   Can you rephrase this?

Reviewer 3: @JustinCappos does this increase clarity?

```suggestion
This component has multiple purposes which are all bound to information available on Kubernetes nodes.
```

Reviewer 2: I don't understand what it means to have "a purpose bound to information"?   What is this trying to say?   Can you avoid the word bound and explain in more detail?

Author: > I don't understand what it means to have "a purpose bound to information"? What is this trying to say? Can you avoid the word bound and explain in more detail?

I wanted to emphasize here that the reason we have a node agent is because only through this sensor can we read and access the information available only on the Kubernetes nodes

Reviewer 2: Okay, your explanation here helps.   Can you work something to this effect into the text so we can resolve the comment?

> This component has multiple purposes which are all bound to information available on Kubernetes nodes.

Reviewer 3: ```suggestion
This component has multiple purposes which are all bound to information available on Kubernetes nodes— It is through this sensor that we can read and access the information available only on the Kubernetes nodes.
```

Author: Thanks @eddie-knight , but we have updated the same line in parallel 😄 WDYT about the current version

* GitHub:  [Discussion](https://github.com/kubescape/kubescape/discussions) or [issues](https://github.com/kubescape/kubescape/issues)

#### Outbound Communication

* Twitter: [@Kubescape](https://x.com/Kubescape)
* Kubescape site: [Kubescape.io](https://kubescape.io/)

##### Community meetings

Kubescape [community meetings](https://zoom.us/j/95174063585) are held on Zoom every other week at 15:00 CET. ([See that in your local time zone](https://time.is/compare/1500_in_CET). :headphones:

Community members are encouraged to join and add to the discussion via the [community meeting notes document](https://docs.google.com/document/d/1X_eyhPzJvb4ascVQ2e0jN87LAvq7lTuXT5d4gQxi8us/edit?usp=sharing) to add an item. You can also see the summary of previous discussions here. 🖥️

## Security issue resolution

For a complete list of closed security issues, please refer to [this link](https://github.com/kubescape/kubescape/issues?q=is%3Aissue+label%3Asecurity+is%3Aclosed)

### Responsible Disclosures Process

> In case of suspected security issues, incidents, or vulnerabilities, both external and internal to the project, Kubescape has a responsible disclosure process in place. The process is designed to handle security vulnerabilities quickly and sometimes privately. The primary goal of this process is to reduce the time users are vulnerable to publicly known exploits.

Reviewer 2: Consider pointing them below for the info about how to report (or just trimming this header).   It makes it sound like they have to fish for reporting info, when you do in fact provide it in a moment.

Author: You're right, I have added a link

In addition, Kubescape has a Fix Disclosure Process that includes disclosing forthcoming fixes to users, completed within 1-7 days of disclosure. The Fix Lead will create a GitHub issue in the Kubescape project to inform users that a security vulnerability has been disclosed and that a fix will be made available, with an estimation of the release date.

## Incident Response

There is a template for incident response for reference [here](https://github.com/cncf/tag-security/blob/main/community/resources/project-resources/templates/incident-response.md)

## Appendix

### Known issues over time

* At the time of preparation, the project had not had a private disclosure of any kind
* Public issues can be found with the `security` label [here](https://github.com/kubescape/kubescape/labels/security)

### OpenSSF best practices

> The Kubescape project is continuously improving its practices based on the OpenSSF recommendations, see scores [here](https://securityscorecards.dev/viewer/?uri=github.com/kubescape/kubescape)

Reviewer 2: Have you thought about a best practices badge?   https://openssf.org/projects/best-practices-badge/   (This is actually what is meant by the requirement.)

Reviewer 3: @slashben 

```suggestion
The Kubescape project is continuously improving its practices based on the OpenSSF recommendations, see [Scorecard Results](https://securityscorecards.dev/viewer/?uri=github.com/kubescape/kubescape) and [Best Practices Badge](https://www.bestpractices.dev/en/projects/6944)
```

# Kubescape Self-assessment

This document is intended to aid the project's maintainers, contributors, and users understand the project's security status and help identify points of improvement.

> *The project seeks a security reviewer for this document who is not part of the project maintainer team or community.*

Reviewer 4: ```suggestion
```

Does it?  This is a self assessment at this point, I would remove this line.

* Kubescape CLI scans pre-deployment and post-deployment configurations and images and provides actionable results

#### Security Goals

* All access to the data produced by Kubescape Operator should be only available to those subjects who have the proper RBAC authorizations
* Kubescape should only share information to מon-Kubernetes endpoints in cases when the user explicitly defined it
* Kubescape should only have access to objects it needs to complete its job
* Triggers should only come from authenticated and authorized sources

### Non-goals

* Kubescape is not meant to enforce access control beyond what the Kubernetes API already provides

## Self-assessment use

This self-assessment is not intended to provide a security audit of Kubescape or function as an independent assessment or attestation of Kubescape's security health.

This document provides Kubescape users with an initial understanding of Kubescape's security, where to find existing security documentation, Kubescape's plans for security, and a general overview of Kubescape's security practices, both for the development of Kubescape and its operational security.

> This document provides the CNCF TAG-Security with an initial understanding of Kubescape to assist in a joint assessment necessary for projects under incubation. Taken together, this document and the joint assessment serve as a cornerstone for if and when Kubescape is preparing for a security audit.

Reviewer 4: Joint assessments are not required for incubation.  This piece of boilerplate is for _incubating_ projects to move to graduation.

```suggestion
```

## Security functions and features

In general, Kubescape is a security project; therefore, we do not list all the features here, only the "non-functional" security features.

### Critical

#### Kubernetes API server authentication

Since Kubescape is mainly driven and uses an API server to store objects, it is critical to validate the Kubernetes API server properly.
This is true for regular API server clients and even more crucial for the *Storage* component, which needs mTLS to function properly to prevent unwanted access.

#### Denial of events

Runtime detection relies heavily on eBPF and container runtime events. Allowing such inputs is critical to maintaining reliable detection capabilities.

#### Masking sensitive results

Since results can contain sensitive information (Kubernetes secrets, environment variables with keys) Kubescape masks these fields from the result objects

> ### Security Relevant

Reviewer 4: > A listing of security relevant components of the project with brief description. These are considered important to enhance the overall security of the project, such as deployment configurations, settings, etc. These should also be included in threat modeling.

On one hand, nothing is security relevant, as it's never in the critical path?
On the other hand, the entire project is security relevant?

I would remove the two sections below, I think they don't add value.

Author: I was advised to keep it like this for now to meet other formatting, but we should work on this later
